tpmutil/mssim/mssim.go

```diff
@@ -0,0 +1,214 @@
+		platformAddr = "127.0.0.1:2322"
+	}
+
+	dial := (&net.Dialer{}).DialContext
+	conn, err := dial(ctx, "tcp", platformAddr)
+	if err != nil {
+		return nil, fmt.Errorf("dial platform address: %v", err)
+	}
+	defer conn.Close()
+
+	// Startup the simulator. This order of commands copies IBM's TPM2 tool's
+	// "powerup" command line tool and will reset the simulator.
+	//
+	// https://sourceforge.net/projects/ibmtpm20tss/
+
+	var buff [4]byte
+	if err := sendPlatformCommand(conn, buff, tpmSignalPowerOff); err != nil {
+		return nil, fmt.Errorf("power off platform command failed: %v", err)
+	}
+	if err := sendPlatformCommand(conn, buff, tpmSignalPowerOn); err != nil {
+		return nil, fmt.Errorf("power on platform command failed: %v", err)
+	}
+	if err := sendPlatformCommand(conn, buff, tpmSignalNVOn); err != nil {
+		return nil, fmt.Errorf("nv on platform command failed: %v", err)
+	}
+
+	// Gracefully close the connection.
+	binary.BigEndian.PutUint32(buff[:], tpmSessionEnd)
+	if _, err := conn.Write(buff[:]); err != nil {
+		return nil, fmt.Errorf("shutdown platform connection failed: %v", err)
+	}
+
+	cmdConn, err := dial(ctx, "tcp", cmdAddr)
+	if err != nil {
+		return nil, fmt.Errorf("dial command address: %v", err)
+	}
+
+	return &Conn{conn: cmdConn}, nil
+}
+
+// sendPlatformCommand sends device management commands to the simulator.
+//
+// See: "D.4.3.2. PlatformServer()"
+func sendPlatformCommand(conn net.Conn, buff [4]byte, u uint32) error {
+	binary.BigEndian.PutUint32(buff[:], u)
+	if _, err := conn.Write(buff[:]); err != nil {
+		return fmt.Errorf("write platform command: %v", err)
+	}
+	if _, err := io.ReadFull(conn, buff[:]); err != nil {
+		return fmt.Errorf("read platform command: %v", err)
+	}
+	rc := binary.BigEndian.Uint32(buff[:])
+	if rc != 0 {
+		return fmt.Errorf("unexpected platform command response code: 0x%x", rc)
+	}
+	return nil
+}
+
+const (
+	connStateIdle    = 0
+	connStateRunning = 1
+)
+
+// Conn is a Microsoft Simulator client that can be used as a connection for the
+// tpm2 package.
+type Conn struct {
+	// Cached connection
+	conn net.Conn
+
+	// Response bytes left over from the previous read.
+	prevRead []byte
+}
+
+// Read a response from the simulator. If the response is longer than the provided
+// buffer, the remainer will be cached for the next read.
+func (c *Conn) Read(b []byte) (int, error) {
+	if c.prevRead != nil {
+		n := copy(b, c.prevRead)
+		if len(c.prevRead) >= len(b) {
+			// If lengths are equal return 0 on the next read to indicate there's
+			// nothing left to read.
+			c.prevRead = c.prevRead[n:]
+		} else {
+			c.prevRead = nil
+		}
+		return n, nil
+	}
+
+	// Response frame:
+	// - uint32 (size of response)
+	// - []byte (response)
+	// - uint32 (always 0)
+	var respLen uint32
+	if err := binary.Read(c.conn, binary.BigEndian, &respLen); err != nil {
+		return 0, fmt.Errorf("read MS simulator response header: %v", err)
+	}
+
+	resp := make([]byte, int(respLen)+4)
```

Comment:
io.ReadFull respLen bytes, and then read rc separately via binary.Read

Reply:
done